### ISSUE TITLE
Syntax tree [src] links all point to macros.rs

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/syn#304]
For example the [src] link from [`syn::Abi`](https://docs.rs/syn/0.12.0/syn/struct.Abi.html) goes [here](https://docs.rs/syn/0.12.0/src/syn/macros.rs.html#47-207) rather than to where you probably want to go in ty.rs.

From experimenting a bit, it seems that the link goes wherever the `pub` token originates from. In this case macros.rs is discarding the `pub` token from the call site and coming up with its on `pub` token, which is why the links go where they do.

```rust
(
    $(#[$attr:meta])*
    pub struct $name:ident $($rest:tt)*
) => {
    $(#[$attr])*
    #[cfg_attr(feature = "extra-traits", derive(Debug, Eq, PartialEq, Hash))]
    #[cfg_attr(feature = "clone-impls", derive(Clone))]
    pub struct $name $($rest)*
};
```

Ideally we could preserve the original `pub` token by matching `$vis:ident struct $name:ident` and emitting `$vis struct $name`, but that triggers a macro_rules ambiguity prior to Rust 1.20.0.

```
error: local ambiguity: multiple parsing options: built-in NTs ident ('vis') or 1 other option.
   --> src/macros.rs:91:19
    |
91  |                   $(#[$variant_attr])*
    |                     ^
```

We may need to go with either a TT-muncher or some dumb syntax that resolves the ambiguity and lets us keep the `pub` token on 1.15.0, like:

```rust
ast_struct! {
    /// ...
    @pub struct Abi {
```

**Comment by dtolnay** [comment]
Filed https://github.com/rust-lang/rust/issues/47817 to consider fixing this in rustdoc.

**Comment by dtolnay** [comment]
Fixed in #376 and 7f2dce7ede0e0d8c973a88f76e61544393be2d3a + e3a6066517dab34ecd6715bf4f9c87b624cf9908.
